### PR TITLE
feat(http): serialize search parameters from request options

### LINKED ISSUE #2417: feat(http): implement support for headers and url params objects in http
The constructors for `Headers` and `UrlSearchParams` have already been implemented to near-spec ([Headers](https://fetch.spec.whatwg.org/#terminology-headers), [UrlSearchParams](https://url.spec.whatwg.org/#urlsearchparams)) in PR #2360, but have not been incorporated into the request process. These should be incorporated in such a way that a dictionary or fully-constructed object of each type can be attached to request options passed to `http`, and will properly be attached to the outgoing request (headers), or serialized into the url string (params).

### PR DESCRIPTION
**Pull request by caitp**
WIP

Closes #2417

### CODE REVIEW

**Comment by jeffbcross** [comment]
@caitp is there anything blocking this PR?

**Comment by caitp** [comment]
The test hasn't been passing yet, so that's one

**Comment by jeffbcross** [comment]
@caitp I meant are you working on this, and is anything blocking you from finishing :)

**Comment by caitp** [comment]
> @caitp I meant are you working on this, and is anything blocking you from finishing :)

I've just pushed a new version which passes each unit test task locally, so this one is probably ready for review.

**Comment by caitp** [comment]
needs info from TypeScript/Dart-knowledgable people

File: modules/angular2/src/http/static_request.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import {
   RegExpWrapper,
   CONST_EXPR,
   isPresent,
-  isJsObject
+  isJsObject,
```
Comment (jeffbcross) [inline modules/angular2/src/http/static_request.ts]: Trailing comma

Comment (jeffbcross) [inline modules/angular2/src/http/static_request.ts]: I'm surprised enforce-format and lint didn't complain about this

File: modules/angular2/src/http/static_request.ts [hunk 2]
```diff
@@ -34,12 +34,18 @@ export class Request {
...
+
   // TODO: support URLSearchParams | FormData | Blob | ArrayBuffer
   private _body: string;
   cache: RequestCacheOpts;
   constructor(requestOptions: RequestOptions) {
     // TODO: assert that url is present
     this.url = requestOptions.url;
+    if (isPresent(requestOptions.search)) {
+      let search = requestOptions.search.toString();
+      if (search.length > 0) this.url = (this.url + '?' + search);
```
Comment (jeffbcross) [inline modules/angular2/src/http/static_request.ts]: If `url` already had query parameters, this would create an invalid url.

Comment (caitp) [inline modules/angular2/src/http/static_request.ts]: Good point, this is also a good test case. Will fix that today.

File: modules/angular2/src/http/base_request_options.ts [hunk 3]
```diff
@@ -56,7 +62,8 @@ export class RequestOptions implements IRequestOptions {
       credentials: isPresent(options) && isPresent(options.credentials) ? options.credentials :
                                                                           this.credentials,
       cache: isPresent(options) && isPresent(options.cache) ? options.cache : this.cache,
-      url: isPresent(options) && isPresent(options.url) ? options.url : this.url
+      url: isPresent(options) && isPresent(options.url) ? options.url : this.url,
+      search: isPresent(options) && isPresent(options.search) ? options.search : this.search
```
Comment (jeffbcross) [inline modules/angular2/src/http/base_request_options.ts]: I'm trying to imagine how I would expect `merge()` to behave if there is a `search` value on the existing and new options objects. For the use case of wanting to have a global or shared set of parameters (like an auth_token for all request to Github), I would expect my request-specific params to be merged with the shared params. I think cases of wanting to outright replace the inherited object with the new object are less common.

Comment (jeffbcross) [inline modules/angular2/src/http/base_request_options.ts]: Could you change the behavior to merge, and add a test that updating a param to null would remove it from the querystring output?

**Comment by jeffbcross** [comment]
Left some comments on requested behavior changes. LMK if you disagree.

File: modules/angular2/src/http/base_request_options.ts (hunk 3 above)
Comment (caitp) [inline modules/angular2/src/http/base_request_options.ts]: I sent some messages on slack about this --- I think if we're going to do that, it would be wise to rename the interface, but I'll let that wait so that the other patch isn't blocked for too long. However, merging rather than replacing is more complicated imho. Are we sure we want to do that?

File: modules/angular2/src/facade/lang.dart [hunk 4]
```diff
@@ -103,6 +103,10 @@ class StringWrapper {
     return s.startsWith(start);
   }
 
+  static endsWith(String s, String end) {
```
Comment (jeffbcross) [inline modules/angular2/src/facade/lang.dart]: Looks like this isn't used anywhere, and is also not implemented in JS facade. Please remove.

Comment (caitp) [inline modules/angular2/src/facade/lang.dart]: there might be a few other bits of cleanup to do, do you see anything that looks more like a "this won't work or doesn't do what we want" kind of thing?

Comment (caitp) [inline modules/angular2/src/facade/lang.dart]: or if you can spot missing test coverage I'll add some more tests before checking in

**Comment by jeffbcross** [comment]
Looks good with some cleanup and squashing.

**Comment by jeffbcross** [comment]
@caitp the merge label is special now. For whitelisted committers, it will put the PR in the presubmit queue as-is. When you're ready to merge it, you can just push to upstream presubmit-caitp-pr-3020
